### ISSUE TITLE
Investigate if store functions should return a `Result`

### ISSUE THREAD

**Issue by plafer** [cosmos/ibc-rs#381]
We currently return early if one of the writes to the store fails (e.g. [here](https://github.com/cosmos/ibc-rs/blob/8ed5c7e1492256805185a9fab1557c4af0f59ff8/crates/ibc/src/core/ics04_channel/context.rs#L331)). ibc-go assumes writes never fail (e.g. [here](https://github.com/cosmos/ibc-go/blob/d4b9f7359f6f407c135e5e71b59262407138281a/modules/core/04-channel/keeper/packet.go#L257)).

We should think whether we want to also assume that writes never fail, or if not, what to do when a write fails.

**Comment by romac** [comment]
Any store backed by some kind of I/O device (file system, network, memory in some specific cases, etc) can and likely will fail at some point. Therefore I believe we should reflect this at the type level and deal with those failures accordingly (eg. by simply bubbling them up to the host application).

What do you think?

**Comment by plafer** [comment]
Right, I thought some more about this since yesterday, and I think it's easily solvable if we document that "if `validate/execute/deliver()` returns an error (any top-level function), all state changes need to be reverted". This is what we do in basecoin-rs, and I think an undocumented assumption that we make everywhere in ibc-rs (which makes sense but I still keep forgetting from time to time). The fix is probably to just clearly document this requirement.
